### PR TITLE
Decouple Happychat UI components

### PR DESCRIPTION
**Pull request by nosolosw**
Master issue #18670

This is to allow happychat-client project to use the UI components without pulling unnecessary Calypso dependencies. Note that at the end of these steps, the components will live in the happychat-client repo - this is just an intermediate step to make progress.

### CODE REVIEW

**Comment by nosolosw** [comment]
Need to fix the emojify tests.

**Comment by nosolosw** [comment]
This PR is only an intermediate step so I can import the UI components from Happychat client and have a steady rhythm while avoiding big-sized PRs. I don't think this should be the final state of UI components.

Perhaps it's worth expanding the process I'm following to extract this from Calypso that I've outlined [here](https://github.com/Automattic/wp-calypso/issues/18670#issuecomment-336087183):

1. Create Calypso-independent pieces. This is already done for the state and this PR aims to do the same for UI components.
2. Import from Calypso those components and use them in Happychat client. I'm working on this now. [See](https://github.com/Automattic/happychat-client/pull/6).
3. Extract and move them from Calypso to Happychat client bit by bit.
4. Profit :tada: 

The external Happychat client API will be consolidated and fleshed out in step 3. Until then, it's general house-cleaning in a way that I hope is reviewer-friendly size-wise.

**Comment by nosolosw** [comment]
Oh, also: the reason I cannot have connected and unconnected components in the same file is that although you use _named imports_, the whole file will be parsed anyway. That means that a file that contains both connected and unconnected components with things like calypso libs, config, selectors, localize, etc is unusable in other projects.

An alternative approach to this PR that I've considered is: creating connected and unconnected components for every piece (composer, timeline, etc) in separate files. The reason I choose this approach instead is that I think it would require less work and provide more clarity to flesh out the external API in the upcoming steps.

File: client/components/emojify/index.jsx [hunk 1]
```diff
@@ -31,11 +31,15 @@ export default class Emojify extends PureComponent {
 		this.parseEmoji();
 	}
 
+	setRef( component ) {
```
Comment (southp) [inline client/components/emojify/index.jsx]: By using the fat arrow syntax here, 
```
setRef = ( component ) => {
```
we should be able to eliminate the `this.setRef.bind( this )` in the constructor.

File: client/components/happychat/index.jsx [hunk 2]
```diff
@@ -54,51 +93,93 @@ export class Happychat extends Component {
...
+	onSendTyping: PropTypes.func,
+	onSetCurrentMessage: PropTypes.func,
 	setBlurred: PropTypes.func,
 	setFocused: PropTypes.func,
+	timeline: PropTypes.array,
+	translate: PropTypes.func,
+	twemojiUrl: PropTypes.string,
+};
+
+const isCurrentUserHelper = currentUser => ( { user_id, source } ) => {
```
Comment (southp) [inline client/components/happychat/index.jsx]: I'd like to propose that we brainstorm better name for this function :)
For me, it was a bit over-generic and I had struggled a bit about why `source === 'customer'` matters if we just want to check if a `user_id` belongs to the current user.

Based on its usage, maybe something like `isMessageFromCurrentUser` would be clearer?

File: client/me/happychat/main.jsx [hunk 3]
```diff
@@ -29,25 +43,102 @@ export class HappychatPage extends Component {
...
+};
+
+const mapState = state => {
+	const currentUser = getCurrentUser( state );
+	return {
+		chatStatus: getHappychatChatStatus( state ),
+		connectionStatus: getHappychatConnectionStatus( state ),
+		currentUserEmail: currentUser.email,
+		disabled: ! canUserSendMessages( state ),
+		isCurrentUser: isCurrentUserHelper( currentUser ), // see redux-no-bound-selectors eslint-rule
```
Comment (southp) [inline client/me/happychat/main.jsx]: The same opinion as above :)

File: client/components/emojify/index.jsx (hunk 1 above)
Comment (nosolosw) [reply client/components/emojify/index.jsx]: I know! But I'm trying to refrain from using `stage2` features.

File: client/me/happychat/main.jsx (hunk 3 above)
Comment (nosolosw) [reply client/me/happychat/main.jsx]: Fixed in 6902e72881ca01dedd3ec88a6289b43ae19acc98

File: client/components/happychat/index.jsx (hunk 2 above)
Comment (nosolosw) [reply client/components/happychat/index.jsx]: Fixed in 6902e72881ca01dedd3ec88a6289b43ae19acc98

File: client/me/happychat/main.jsx (hunk 3 above)
Comment (nosolosw) [reply client/me/happychat/main.jsx]: Fixed in 6902e72881ca01dedd3ec88a6289b43ae19acc98

**Comment by nosolosw** [comment]
@southp implemented your suggestions.

Regarding your questions:

- I don't think this should be the final state of UI components, and I agree that there is some common code between `index` and `main` that can be factored out. But I prefer to do that in upcoming steps, as the rationale for this PR is do the Minimal Viable Thing that enables us to move the "pure" happychat UI components from Calypso to Happychat client repository ASAP. With that in place, we can iterate the API surface without falling into (too many) integration issues that drag a lot of time.

- HappychatConnection is used in a lot of places independently from the Happychat form and panel, so, at this point, we need the connected component.

File: client/me/happychat/main.jsx (hunk 3 above)
Comment (nosolosw) [reply client/me/happychat/main.jsx]: Fixed in 6902e72

File: client/components/happychat/index.jsx (hunk 2 above)
Comment (nosolosw) [reply client/components/happychat/index.jsx]: Fixed in 6902e72

**Comment by nosolosw** [comment]
Rebased to fix some conflicts with `master`.

**Pull request merged**